### PR TITLE
Refactor [2/N]

### PR DESCRIPTION
**Pull request by t-young31**
Continues #63 

I'd strongly recommend we add some more comprehensive system tests before merging this.

***
### Guidelines for reviewers
These commits have been reviewed as separate PRs on my fork previously. If this PR warrants further review I'd strongly recommend reviewing the individual commits, with the exception of the one that just clang-formats the entire `iterator.cpp` file.

***
### Extensions

- Convert the cerr 'prints' to spdlog logging calls
- Cary on the refactor past the start of the main loop
- Change the non-matlab arrays (defined in `arrays.h`) to something more robust
- Use the defined `.clang-format` file to reformat the entire code base (ideally as a precommit hook)
- Add _many_ more unit tests (this refactor largely focused on extracting classes and structs that have matlab-dependent constructors, thus didn't warrant testing, if they're going to be rewritten)

***
### Dependencies

- #64 
- #30

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Generated from CLion C/C++ Code Style settings
```
Comment (samcunliffe) [inline .clang-format:1]: Did you mean to commit this?
We will have a unified clang-format settings, so it's not a bad idea IMO.

**Review by samcunliffe — APPROVED** [review]
Just checking/reminding myself of what's going on here (and double checking for anything I might've missed). 

→ I've already reviewed all of the changes in little chunks: so approving here too.

File: .clang-format (hunk 1 above)
Comment (t-young31) [reply .clang-format:1]: yep, but feel free to remove the comment/modify accordingly. strong in favour of clang-formatting the whole code base

Comment (samcunliffe) [reply .clang-format:1]: Ah, I meant the whole file. A common clang-format config is a good idea then.

**Review by prmunro — APPROVED** [review]

File: tdms/src/arrays.cpp [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+  }
+
+  matrix = cast_matlab_2D_array(mxGetPr(ptr), n_rows, n_cols);
+  this->n_cols = n_cols;
+  this->n_rows = n_rows;
+}
+
+Pupil::~Pupil() {
+  free_cast_matlab_2D_array(matrix);
+  matrix = nullptr;
```
Comment (t-young31) [inline tdms/src/arrays.cpp:194]: this is an appalling hack – please can we get rid of matlab soon

**Review by samcunliffe — APPROVED** [review]

**Pull request merged**